### PR TITLE
make type definitons `"module": "nodenext"` compatible

### PR DESCRIPTION
**Pull request by wight554**
* see https://github.com/microsoft/TypeScript/issues/48845

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by climba03003 — DISMISSED** [review]
Not sure which TypeScript update make it not work.
But it will break more people than it should.
So, I think this PR should not be merged.

**Comment by wight554** [comment]
> Not sure which TypeScript update make it not work. But it will break more people than it should. So, I think this PR should not be merged.

Can you give more context on "breaking"? It passes the test suite, so everything should be fine
Also codebase itself is untouched

**Comment by climba03003** [comment]
> > Not sure which TypeScript update make it not work. But it will break more people than it should. So, I think this PR should not be merged.
> 
> Can you give more context on "breaking"? It passes the test suite, so everything should be fine Also codebase itself is untouched

I would like to see a repro for this error first.

```ts
// tsconfig.json
{
  "compilerOptions": {
    "module": "NodeNext"
  }
}

// index.ts
import defaultFastifyCookie, { fastifyCookie } from "@fastify/cookie";
import Fastify from "fastify";

const fastify = Fastify();
fastify.register(fastifyCookie);
fastify.register(defaultFastifyCookie);

// package.json
{
  "name": "test",
  "version": "1.0.0",
  "scripts": {},
  "dependencies": {
    "@fastify/cookie": "^6.0.0",
    "@types/node": "^17.0.35",
    "fastify": "^4.0.0-rc.2",
    "typescript": "^4.7.2"
  }
}
```

From what I have tested, I do not face any error currently.
image
image
It is working correctly.

**Comment by wight554** [comment]
> > > Not sure which TypeScript update make it not work. But it will break more people than it should. So, I think this PR should not be merged.
> > 
> > 
> > Can you give more context on "breaking"? It passes the test suite, so everything should be fine Also codebase itself is untouched
> 
> I would like to see a repro for this error first.
> 
> ```ts
> // tsconfig.json
> {
>   "compilerOptions": {
>     "module": "NodeNext"
>   }
> }
> 
> // index.ts
> import defaultFastifyCookie, { fastifyCookie } from "@fastify/cookie";
> import Fastify from "fastify";
> 
> const fastify = Fastify();
> fastify.register(fastifyCookie);
> fastify.register(defaultFastifyCookie);
> 
> // package.json
> {
>   "name": "test",
>   "version": "1.0.0",
>   "scripts": {},
>   "dependencies": {
>     "@fastify/cookie": "^6.0.0",
>     "@types/node": "^17.0.35",
>     "fastify": "^4.0.0-rc.2",
>     "typescript": "^4.7.2"
>   }
> }
> ```
> 
> From what I have tested, I do not face any error currently. image image It is working correctly.

sure
a bit more than this:
```
{
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
 }
```

**Comment by wight554** [comment]
also `"type": "module",` in `package.json`
if you investigate further, you'll see that using default import breaks typescript (not callable) and using named import breaks node runtime after successful compilation:
```
SyntaxError: Named export 'fastifyCookie' not found. The requested module 'fastify-cookie' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:
```

**Comment by climba03003** [comment]
The problem actually comes from 
`"type": "module"` with `"module": "NodeNext"`

**Comment by wight554** [comment]
it seems to be expected behaviour from TS side, it is CJS module so it should use CJS export
you can find more investigation in this issue: https://github.com/bendrucker/snakecase-keys/pull/72

File: plugin.d.ts [hunk 1]
```diff
@@ -84,33 +84,38 @@ declare module 'fastify' {
...
+  export const fastifyCookie: FastifyPluginCallback<NonNullable<FastifyCookieOptions>>;
 
-export interface FastifyCookieOptions {
-  secret?: string | string[] | Signer;
-  parseOptions?: CookieSerializeOptions;
+  export { fastifyCookie as default };
 }
 
-declare const fastifyCookie: FastifyPluginCallback<NonNullable<FastifyCookieOptions>>;
+declare function fastifyCookie(): FastifyPluginCallback<NonNullable<fastifyCookie.FastifyCookieOptions>>;
```
Comment (climba03003) [inline plugin.d.ts:119]: I don't think it is the correct type.

```ts
import defaultFastifyCookie, { fastifyCookie } from "@fastify/cookie";

typeof defaultFastifyCookie === 'function'
typeof defaultFastifyCookie() === 'function'

const isFunc1: Function = defaultFastifyCookie
const isFunc2: Function = defaultFastifyCookie() // it should fail
const isFunc3: Function = fastifyCookie
```

Comment (wight554) [reply plugin.d.ts:119]: yeah, correct point
seems like TS still lacks a way to specify type for function declarations:
https://github.com/microsoft/TypeScript/issues/22063
I'll try to workaround this

**Comment by wight554** [comment]
@climba03003 pushed changes to fix default export function type

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you please add a unit test? we use tsd to test our types

**Comment by wight554** [comment]
> Can you please add a unit test? we use tsd to test our types

Idk what can also be tested, current test suite looks completed to me

**Review by climba03003 — COMMENTED** [review]
I will not block this PR to be merged and I would like to see the namespace written in `PascalCase`.

@fastify/typescript just ping more people in to decide what's the next steps.

**Comment by wight554** [comment]
> I will not block this PR to be merged and I would like to see the namespace written in `PascalCase`.
> 
> @fastify/typescript just ping more people in to decide what's the next steps.

PascalCase sounds good to me, will add corresponding changes

**Comment by wight554** [comment]
Ah well actually @climba03003 PascalCase is no go here. fastifyCookie namespace is merged with fastifyCookie function, and PascalCase for function name doesn't sound great

**Comment by climba03003** [comment]
> PascalCase is no go here

Sad for this info.

File: plugin.d.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 /// <reference types='node' />
 
-import { FastifyPluginCallback } from 'fastify';
+import { FastifyPluginCallback } from "fastify";
 
-declare module 'fastify' {
+declare module "fastify" {
```
Comment (RafaelGSS) [inline plugin.d.ts:5]: can you restore it? It seems unrelated to the PR goal.

**Review by RafaelGSS — COMMENTED** [review]
Could you include a test for ESM? Reference: https://github.com/fastify/fastify/tree/main/test/esm

**Comment by wight554** [comment]
> Could you include a test for ESM? Reference: https://github.com/fastify/fastify/tree/main/test/esm

I don't really see any reason to test esm since module is CJS only, these changes make typescript read module as CJS

**Comment by RafaelGSS** [comment]
> > Could you include a test for ESM? Reference: https://github.com/fastify/fastify/tree/main/test/esm
> 
> I don't really see any reason to test esm since module is CJS only, these changes make typescript read module as CJS

If the current test suite was any good, it should have detected this problem, shouldn't it? I'm just concerned about preventing a potential regression in any further PR.

**Comment by wight554** [comment]
> > > Could you include a test for ESM? Reference: https://github.com/fastify/fastify/tree/main/test/esm
> > 
> > 
> > I don't really see any reason to test esm since module is CJS only, these changes make typescript read module as CJS
> 
> If the current test suite was any good, it should have detected this problem, shouldn't it? I'm just concerned about preventing a potential regression in any further PR.

I think there is no regression as long as it passes current tsd suite, not sure what I can do to test esm in this module since there is no esm (.mjs) module exported

**Comment by wight554** [comment]
Actually there's a regression already since https://github.com/fastify/fastify-cookie/commit/8da8fbe3fa72bde19e066b1c49ec902dc7c8f06f was added
types don't expect [these lines](https://github.com/fastify/fastify-cookie/blob/master/plugin.js#L118...##L121) to be inside of fastifyCookie named export

File: plugin.d.ts (hunk 2 above)
Comment (wight554) [reply plugin.d.ts:5]: tbh I think it's fine to keep those, that's was formatter does for me every time I touch that file, also makes it consistent with other plugins (at least fastify-static)

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@wight554 could you rebase?
@RafaelGSS ptal

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Comment by wight554** [comment]
> @wight554 could you rebase? @RafaelGSS ptal

should be g2g
